Restructure significant part of the controller documentation (#12549)
nextcloud/documentation: developer_manual/basics/controllers.rst

To use OCS in your API you can use the **OCP\\AppFramework\\OCSController** base class and return your data in the form of a **DataResponse** in the following way:

.. code-block:: php

    <?php
    namespace OCA\MyApp\Controller;

    use OCP\AppFramework\Http\DataResponse;
    use OCP\AppFramework\Http\Attribute\NoAdminRequired;
    use OCP\AppFramework\OCSController;

    class ShareController extends OCSController {

        #[NoAdminRequired]
        public function getShares(): DataResponse {
            return new DataResponse([
                //Your data here
            ]);
        }

    }

For ``OCSController`` classes and their methods, :ref:`responders <controller-responders>` can be registered as with any other ``Controller`` method.
The ``OCSController`` class have however automatically two responders pre-installed:
Both JSON (``application/json``) and XML (``text/xml``) are generated on-the-fly depending on the request by the browser/user.
To select the output format, the ``?format=`` query parameter or the ``Accept`` header of the request work out of the box, no intervention is required.

> It is advised to prefer the header generally, as this is the more programmatic way.

Reviewer: ```suggestion
It is advised to prefer the header generally, as this is the more standardized way.
```

The `?format=` is pretty bad IMO, since it does not adhere to HTTP standards while the `Accept` header is exactly made for this purpose. I'm not even sure we should explain here that such a query parameter ever existed.

Author: I'd keep it for now. n this big PR, we have enough stuff to think about. The deletion of the `?format=` parameter can be postponed (as it is currently part of the docs) into a separate PR. Do you agree?

Author: The sugestion is resolved, just keeping this open about the `?format=` question

    }

For ``OCSController`` classes and their methods, :ref:`responders <controller-responders>` can be registered as with any other ``Controller`` method.
The ``OCSController`` class have however automatically two responders pre-installed:
Both JSON (``application/json``) and XML (``text/xml``) are generated on-the-fly depending on the request by the browser/user.
To select the output format, the ``?format=`` query parameter or the ``Accept`` header of the request work out of the box, no intervention is required.
It is advised to prefer the header generally, as this is the more programmatic way.

> In order to make routing work for OCS routes you need to add :ref:`a separate 'ocs' entry<routes_ocs>` to the routing table in ``appinfo/routes.php`` of your app.

Reviewer: This should also mention the attributes you can use since a few versions and link to https://github.com/nextcloud/documentation/blob/de2b54eea51e510b55ff1e1d7d22f5f168eea5f3/developer_manual/basics/routing.rst#routing as an explanation of how to use both.

Reviewer: Ok I saw you changed the routing explanation in a later commit.
I would still like to have attributes mentioned more clearly though.

Author: How about something like this?

```suggestion
To make routing work for OCS, the route must be registered in the core.
This can be made in two ways:
You can add an attribute `#[ApiRoute]` to the controller method.
Alternatively, you can add :ref:`a separate 'ocs' entry<routes_ocs>` to the routing table in ``appinfo/routes.php`` of your app.
```

Reviewer: `This can be done in two ways` but other than that sounds good.

    * ``@NoTwoFactorRequired``` instead of ``#[NoTwoFactorRequired]``
    * ``@NoCSRFRequired``` instead of ``#[NoCSRFRequired]``

In the following some examples of configurations are given.

Showing an HTML page by the user
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A typical app needs an ``index.html`` page to show all content within.
This page should be visible by all users in the instance.
Therefore, you need to loosen the restriction from admins only (``#[NoAdminRequired]``).

> Additionally, as the user might not have a CSRF checker cookie set yet, the CSRF checks should be disabled (which is fine as this is a template response).

Reviewer: This is true for all pages that the user visits directly via the browser, not only template responses.
I'm not sure how to word it, such that it's not understood as "just disable it everywhere when something doesn't work".

Author: These are just some (three) examples. We could, however, add a sentence that for the sake of the example, the typical case of a template response is used. At least in NC all HTML pages are somewhat templates AFAIK unless you use lazy rendering or such. Am I wrong here?

Reviewer: Right, maybe I was just thinking too much. You can serve custom HTML, but indeed almost everything is template based.

        #[NoAdminRequired]
        public function getShares(): DataResponse {
            return new DataResponse([
                // Your data here
            ]);
        }

    }

> The ``#[NoAdminRequired]`` is needed here as normal users should be able to access the data in fact.

Reviewer: ```suggestion
The ``#[NoAdminRequired]`` is needed here as normal users should be able to access the data.
```
